CORE-9875: Topic ids: add `topic_table::get_name_by_id(...)`

pgellert opened the pull request:
This PR implements an efficient lookup of topic names by topic id on the `topic_table`. This will be used by kafka handlers to translate the topic id received on a request to its topic name.

This is implemented as a derived map from topic id to topic name that is updated on every topic modification command.

Fixes https://redpandadata.atlassian.net/browse/CORE-9875

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
cluster: add `topic_table::get_name_by_id(...)`

Implement an efficient lookup for finding a topic by its topic id. This
will be used by kafka handlers to translate the topic id received on a
request to its topic name.

The optional topic id is always checked for nullopt-ness, since the code
needs to work during the migration period when not all topics have topic
ids assigned.

This is implemented using a wrapper class `underlying_map` as well as by
making various fields of `topic_metadata_item` private to provide safer
access to the value type.

The goal is to prevent accidental mutation of the `tp_ns` and `tp_id`
inside the `topic_metadata_item` as much as possible, since that would
violate the invariants of the `by_tp` and `by_id` maps.

pgellert committed:
cluster/tests: test topic id mappings

Add unit tests to validate the topic id mapping behaviour.

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -269,6 +269,9 @@ class topic_table {
       model::topic_namespace_hash,
       model::topic_namespace_eq>;
 
+    using topic_id_mapping_t
+      = chunked_hash_map<model::topic_id, model::topic_namespace>;
```
I think it's worth encapsulating the maps, a starting point might look something like:
```c++
    using underlying_t = chunked_hash_map<
      model::topic_namespace,
      topic_metadata_item,
      model::topic_namespace_hash,
      model::topic_namespace_eq>;

    using topic_id_mapping_t
      = chunked_hash_map<model::topic_id, model::topic_namespace>;

    class underlying_map {
    public:
        const auto& by_tp() const { return _by_tp; }
        // can wrap const versions of find, contains, end, etc,
        // mutable versions might require more care.

        template<typename... Args>
        auto insert(Args&&... args) {
            auto r = _by_tp.insert(std::forward<Args>(args)...);
            if (r.second()) {
                _by_id.insert_or_assign(
                  r.first->second.get_configuration().tp_id, r.first->first);
            }
            return r;
        }

        template<typename... Args>
        auto erase(Args&&... args) {
            auto it = _by_tp.find(std::forward<Args>(args)...);
            if (it != _by_tp.end()) {
                _by_id.erase(it->second.get_configuration().tp_id);
            }
            return _by_tp.erase(it);
        }

        std::optional<model::topic_namespace>
        get_name(model::topic_id tp_id) const {
            if (auto it = _by_id.find(tp_id); it != _by_id.end()) {
                return it->second;
            }
            return std::nullopt;
        }

    private:
        underlying_t _by_tp;
        topic_id_mapping_t _by_id;
    };

    underlying_map _topics;
 ```

BenPope replied:
If you're feeling particularly confident, you could store a `topic_namespace_view`, but at that point an iterator into the other map would be equivalent.

pgellert replied:
Yeah, good idea, I've adopted this now. I still store a `model::topic_namespace` though to keep things simpler.

aanthony-rp reviewed (dismissed):
straightforward interface, well done.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/6b72bd543951d295df39f8b742f37ae3e0d8a4fe..c56a6480131474c748a54d98a23a950aa5b34d18): address code review feedback

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -251,6 +252,83 @@ class topic_table {
+        // mutation
+        // Prefer the mutable interface of underlying_map instead (emplace,
+        // erase, assign_id)
+        auto& by_tp_mutable() { return _by_tp; }
```
Sorry, I'm not a fan of handing out a mutable underlying map.

I'm less concerned with mutable iterators (and we can use deducing this now!)
```suggestion
        template<typename Self>
        constexpr auto begin(this Self&& self) {
            return std::forward<Self>(self)._by_tp.begin();
        }

        template<typename Self>
        constexpr auto end(this Self&& self) {
            return std::forward<Self>(self)._by_tp.end();
        }

        template<typename Self>
        auto find(this Self&& self, model::topic_namespace_view tp) {
            return std::forward<Self>(self)._by_tp.find(tp);
        }
```

BenPope replied:
Actually this doesn't work!

All of this is to avoid things which look a bit scary such as [this line](https://github.com/redpanda-data/redpanda/blob/1f60fc53d971364377ca62ee7084d7943f40c2ce/src/v/cluster/topic_table.cc#L1569)
```c++
                // The topic was re-created, delete and add it anew.
                co_await applier.delete_topic(ns_tp, md_item);
                md_item = co_await applier.create_topic(ns_tp, topic_snapshot);
```

pgellert replied:
I've reworked this now with a `mutate` method

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -251,6 +252,83 @@ class topic_table {
+            if (it != _by_tp.end()) {
+                it->second.get_configuration().tp_id = tp_id;
+                if (tp_id) {
+                    _by_id.emplace(*tp_id, tp);
```
Unlikely but:
```suggestion
                    _by_id.erase(*tp_id);
                    _by_id.emplace(*tp_id, tp);

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -251,6 +252,83 @@ class topic_table {
+          model::topic_namespace tp, std::optional<model::topic_id> tp_id) {
+            auto it = _by_tp.find(tp);
+            if (it != _by_tp.end()) {
+                it->second.get_configuration().tp_id = tp_id;
+                if (tp_id) {
+                    _by_id.emplace(*tp_id, tp);
```
This shouldn't happen, but is possible, I guess:
```suggestion
                auto old_id = it->second.get_configuration().tp_id;
                it->second.get_configuration().tp_id = tp_id;
                if (tp_id) {
                    if (old_id) {
                        _by_id.erase(*old_id);
                    }
                    _by_id.emplace(*tp_id, tp);
```

pgellert replied:
Done

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5db7b50c-0923-4735-a350-ffafe213c42f-->
<!--TEST RESULTS-->
<!--4b4c93a3-8909-40ac-b61e-c6b644752e99-->
<!--TEST RESULTS-->
<!--ffda1932-846e-40ec-9308-7f7a8ecd7987-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65698">build#65698</a></summary>


| test_class               | test_method                       | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:----------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_working_node | {"delete_topic": true, "tick_interval": 5000}                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65698#0196b082-9d58-4d3c-b12b-8f986a55c444 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionBalancerTest    | test_unavailable_nodes            |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65698#0196b082-b621-44f7-8398-bd9802301b9e | FLAKY         | 15/21    | upstream reliability is '74.52667814113597'. current run reliability is '71.42857142857143'. drift is 3.09811 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65698#0196b082-b623-4390-b7e8-ee84ce9279c4 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65981">build#65981</a></summary>


| test_class               | test_method                | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations       | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65981#0196ceae-a557-4188-991a-9fbd83bec4dd | FLAKY         | 20/21    | upstream reliability is '95.52715654952077'. current run reliability is '95.23809523809523'. drift is 0.28906 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations       | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": true}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65981#0196ceae-a558-4c17-a739-7d0a3ebfd30d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| OIDCReauthTest           | test_oidc_reauth           |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65981#0196cead-a976-460e-aa0b-202d1cd17ab1 | FLAKY         | 20/21    | upstream reliability is '99.53917050691244'. current run reliability is '95.23809523809523'. drift is 4.30108 and the allowed drift is set to 50. The test should PASS |
| ScalingUpTest            | test_on_demand_rebalancing | {"partition_count": 1}                                                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65981#0196ceae-a557-4188-991a-9fbd83bec4dd | FLAKY         | 20/21    | upstream reliability is '99.53051643192488'. current run reliability is '95.23809523809523'. drift is 4.29242 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66097">build#66097</a></summary>


| test_class                 | test_method                                       | test_arguments                                                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:--------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RestCatalogConnectionTest  | test_redpanda_connection_to_rest_catalog          | {"cloud_storage_type": 1}                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66097#0196d931-42e7-41af-9618-7b251474fefe | FLAKY         | 20/21    | upstream reliability is '97.70992366412213'. current run reliability is '95.23809523809523'. drift is 2.47183 and the allowed drift is set to 50. The test should PASS |
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments |                                                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66097#0196d931-42e8-4899-8d2b-8efad45bf105 | FLAKY         | 17/21    | upstream reliability is '90.45801526717557'. current run reliability is '80.95238095238095'. drift is 9.50563 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66097#0196d933-e987-414a-a3ea-27c51092021f | FLAKY         | 20/21    |                                                                                                                                                                        |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": true}   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66097#0196d933-e987-414a-a3ea-27c51092021f | FLAKY         | 19/20    |                                                                                                                                                                        |
| DisablingPartitionsTest    | test_disable                                      |                                                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66097#0196d931-42e7-4ed8-ba43-50de20d394ce | FLAKY         | 17/21    | upstream reliability is '87.6923076923077'. current run reliability is '80.95238095238095'. drift is 6.73993 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

dotnwat commented on src/v/cluster/topic_table.h:
```diff
@@ -251,6 +252,83 @@ class topic_table {
+                      it->first);
+                }
+            }
+            return _by_tp.erase(it);
```
should this erase go into the previous scope?

pgellert replied:
It doesn't matter because `_by_tp.erase(_by_tp.end())` is UB anyway. So it's simpler to just end the function with the wrapped method.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/c56a6480131474c748a54d98a23a950aa5b34d18..a9fb8d56cffa09c3d1b83db80fb16f12b203dafb): rework the underlying map to use a `mutate()` method for mutating instead of `by_tp_mutable()`

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -251,6 +253,105 @@ class topic_table {
+                    _by_id.erase(*old_id);
+                }
+                if (new_id) {
+                    _by_id.emplace(*new_id, it->first);
```
What if the new id already exists, but points to another topic name?

Should there also be protection against mutating the topic name?

pgellert replied:
This is the same question as https://github.com/redpanda-data/redpanda/pull/26069#discussion_r2084209360, right?

BenPope commented on src/v/cluster/tests/topic_table_test.cc:
```diff
@@ -599,6 +613,50 @@ FIXTURE_TEST(test_topic_with_schema_id_validation_ops, topic_table_fixture) {
+    BOOST_REQUIRE_EQUAL(topics.get_name_by_id(tp_1_new_id), tp_ns1);
+    BOOST_REQUIRE_EQUAL(topics.get_name_by_id(*tp_id2), tp_ns2);
+}
+
```
I think some more tests are required:
```c++
FIXTURE_TEST(test_topic_mutate, topic_table_fixture) {
    const auto emplace = [this](auto& topics, const ss::sstring& name) {
        auto cmd = make_create_topic_cmd(name, 1, 3);
        auto key = cmd.value.cfg.tp_ns;
        auto value = cluster::topic_table::topic_metadata_item{
          .metadata = cluster::topic_metadata{
            std::move(cmd.value), model::revision_id{0}}};
        return topics.emplace(key, std::move(value));
    };

    {
        // duplicate topic id
        cluster::topic_table::underlying_map topics;

        auto t1_tp_ns = emplace(topics, "topic_1").first->first;
        auto t2_tp_id = emplace(topics, "topic_2")
                          .first->second.metadata.get_configuration()
                          .tp_id;

        topics.mutate(topics.find(t1_tp_ns), [&](auto& t) {
            t.metadata.get_configuration().tp_id = t2_tp_id;
        });

        BOOST_REQUIRE_EQUAL(topics.by_tp().size(), topics.by_id().size());
    }

    {
        // duplicate topic name
        cluster::topic_table::underlying_map topics;

        auto t1_tp_ns = emplace(topics, "topic_1").first->first;
        auto t2_tp_ns = emplace(topics, "topic_2").first->first;

        topics.mutate(topics.find(t1_tp_ns), [&](auto& t) {
            t.metadata.get_configuration().tp_ns = t2_tp_ns;
        });

        BOOST_REQUIRE_EQUAL(topics.by_tp().size(), topics.by_id().size());
    }
}
```

BenPope replied:
Of course, more assertions about the final state of the maps should be made

pgellert replied:
What behaviour would you expect here? These test cases obviously corrupt the `underlying_map` and seem like examples of caller bugs to me rather than test cases we would want to maintain. Perhaps I could add two `vassert`s into `mutate` for this, to ensure that the new topic id doesn't already exist and that the topic name is unchanged?

BenPope replied:
> What behaviour would you expect here?

That the maps are consistent. 

> seem like examples of caller bugs to me rather than test cases we would want to maintain.

Right, but the point of encapsulating the state is to maintain its invariants.

pgellert replied:
I don't think it's obvious how we can make the maps consistent here. I think we should `vassert` here if the invariants are violated and let the caller side make sure that the invariants are upheld because it is clearly a bug if the caller mutates the topic name in-place here.

We could choose to have an `underlying_map` implementation that, in this case, fixes the backing maps (eg. delete the topics at the new topic name or new topic id, reinsert the topic metadata at the new topic name and new topic id). I don't think that behaviour would be useful. That would propagate a bug to some other location in the code and would increase the complexity of the code without corresponding expected behaviour.

BenPope replied:
> That would propagate a bug to some other location in the code

`mutate` could return `std::pair<const_iter, bool>`  so that the error can be handled.

Or, an alternative to mutate, is to have `topic_metadata_item` encapsulate `topic_metadata` a bit more, so that the `tp_id` and `tp_ns` (the "key") are only provided via `const&`, everything else is "value" and is allowed to be mutated in-place. Updating `tp_ns` or `tp_id`, could be done via `underlying_map::extract`, which is probably a bit more standard-like.

pgellert replied:
> mutate could return std::pair<const_iter, bool> so that the error can be handled.

None of the call sites expect to change the topic name, therefore, they would all just ignore/vassert the return value. That seems equivalent to vasserting inside `mutate`.

> Or, an alternative to mutate, is to have topic_metadata_item encapsulate topic_metadata a bit more, so that the tp_id and tp_ns (the "key") are only provided via const&, everything else is "value" and is allowed to be mutated in-place. Updating tp_ns or tp_id, could be done via underlying_map::extract, which is probably a bit more standard-like.

That seems like a bigger change to figure out how that would work on all construction and access sites, and how that would work with serde. I do not think that this level of encapsulation is necessary for the correctness (and future correctness) of this code. I have already improved the encapsulation here significantly compared to what it was before the PR through the `underlying_map`. I think if we want to do further refactoring here, that should be done as a separate ticket and PR from this one.

BenPope replied:
Something like this adds a bit more encapsulation: https://github.com/BenPope/redpanda/commit/eff35271f830229e2c4e7d85c3e88e2cd3e2a800

pgellert replied:
Done, I took your suggestions to reduce the need for `mutate` and added vasserts into `emplace` and `mutate` to assert that they do not lead to a corrupted state. The tests above would now both fail on vasserts.

I've also added some more validation (`errc::topic_id_already_exists`) to check if the input commands (`create_topic_cmd`/`update_topic_properties_cmd`) were to lead to duplicated topic ids.

BenPope replied:
> I've also added some more validation (errc::topic_id_already_exists) to check if the input commands (create_topic_cmd/update_topic_properties_cmd) were to lead to duplicated topic ids.

Should there be tests for that?

pgellert replied:
Yep, added some tests now.

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -251,6 +253,105 @@ class topic_table {
+        }
+
+        template<typename... Args>
+        auto emplace(Args&&... args) {
```
For safety, this should probably return a ` std::pair<const_iterator, bool>`

pgellert replied:
I'm not following why. Could you elaborate please?

BenPope replied:
To prevent:
```c++
model::topic key{"foo"};
_topics.emplace(key, val).first->second.get_configuration().tp_ns.tp = model::topic{"bar"};
```

BenPope replied:
Actually, I guess that we don't even want `emplace(key, val)`, we want:
```c++
auto emplace(auto val&& v) {
   auto key = v.get_configuration().tp_ns;
   _impl.emplace(key, std::move(v));
}

pgellert replied:
Done

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -251,6 +253,105 @@ class topic_table {
+
+        using mut_fn
+          = std::function<void(underlying_t::value_type::second_type&)>;
+        bool mutate(const underlying_t::const_iterator it, const mut_fn& func) {
```
We don't really need the type-erasure here:
```suggestion
        template<
          std::invocable<typename underlying_t::value_type::second_type&> Func>
        bool
        mutate(const typename underlying_t::const_iterator it, Func&& func) {
```

pgellert replied:
Done

BenPope commented on src/v/cluster/topic_table.cc:
```diff
@@ -1527,7 +1561,13 @@ class topic_table::snapshot_applier {
+              partition,
+              ret.get_assignments(),
+              ret.partitions,
+              false);
```
nitpick: It looks like `add_ntp` didn't need to change.

pgellert replied:
Reverted to the old version of `add_ntp`

BenPope commented on src/v/cluster/topic_table.cc:
```diff
@@ -1562,7 +1602,11 @@ ss::future<> topic_table::apply_snapshot(
-                md_item = co_await applier.create_topic(ns_tp, topic_snapshot);
+                auto new_md = co_await applier.create_topic(
+                  ns_tp, topic_snapshot);
+                _topics.mutate(old_it, [&](auto& md_item_mut) {
```
Not a new problem, but I don't see anything protecting `old_it` from invalidation after an async yield.

pgellert replied:
Yeah, that does look both preexisting and scary. My first guess is that this is safe because the controller stm ensures that nothing will be concurrent with `apply_snapshot`, but it's difficult to prove that from the code. I can file a ticket to dig deeper on this if you think that's worth it.

pgellert replied:
All of this is guarded by the `_apply_mtx` of the controller state machine, so there is no concurrency here: https://github.com/redpanda-data/redpanda/blob/aa7cf199fa401375f529c03eb394850ca620e031/src/v/raft/mux_state_machine.h#L225-L227

pgellert committed:
cluster: add `errc::topic_id_already_exists`

This will be returned from command apply methods on topic creation and
topic id assignment if the given topic id already exists.

This error should not really happen (assuming the uuids used for topic
ids are always unique), but it is safer to handle this as an internal
error instead of asserting that they never happen.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/a9fb8d56cffa09c3d1b83db80fb16f12b203dafb..2e512101851b114eb949ca48061153ddb2596e5e) (twice): 
- reworked the PR now with @BenPope's suggestion from https://github.com/redpanda-data/redpanda/pull/26069#discussion_r2086602111
- added `errc::topic_id_already_exists` to add more validation on duplicate topic ids to treat them as an internal error instead of asserting on it

BenPope commented on src/v/cluster/topic_table.h:
```diff
@@ -236,21 +243,170 @@ class topic_table {
+            return _by_tp.contains(std::forward<Args>(args)...);
+        }
+
+        auto emplace(topic_metadata_item val) {
```
nitpick:
```suggestion
        std::pair<underlying_t::const_iterator, bool>
        emplace(topic_metadata_item val) {
```

BenPope reviewed (commented):
Looks great

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/2e512101851b114eb949ca48061153ddb2596e5e..6d7c0ab1e312af5a13f8a7cf4b1d874fe2a1d9b7): add tests for `errc::topic_id_already_exists`

dotnwat reviewed (commented):
let's get replication team review cc @mmaslankaprv @ztlpn

mmaslankaprv commented on src/v/cluster/topic_table.h:
```diff
@@ -236,21 +243,170 @@ class topic_table {
+        }
+
+        template<typename... Args>
+        auto contains(Args&&... args) const {
```
is variadic template required here ?

pgellert replied:
Not required, both `contains` and `find` could just take a `model::topic_namespace_view key`, but the variadic template makes it clear that this just wraps the identically-called methods of `_by_tp`. I'm happy to change to taking `model::topic_namespace_view key` if we're worried about compile times here.

pgellert replied:
Fixed in https://github.com/redpanda-data/redpanda/pull/26069#issuecomment-2886449970

mmaslankaprv commented on src/v/cluster/topic_table.h:
```diff
@@ -236,21 +243,170 @@ class topic_table {
+        auto empty() const { return _by_tp.empty(); }
+
+        template<typename Self, typename... Args>
+        auto find(this Self&& self, Args&&... args) {
```
also here do we need variadic template ?

pgellert replied:
I changed this now in https://github.com/redpanda-data/redpanda/pull/26069#issuecomment-2886449970

mmaslankaprv commented:
lgtm, just one question

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/6d7c0ab1e312af5a13f8a7cf4b1d874fe2a1d9b7..4ab60652c5e1127302b0656c607deb8d7e3949a5): address feedback (avoid variadic templates)
